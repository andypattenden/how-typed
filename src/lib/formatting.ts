import chalk from 'chalk'
import { formattingColours } from '../constants.js'
import type { FormattingColours } from '../@types/index.js'
import { ratings, type Rating } from './ratings.js'

export function formatString(
  string: string,
  fgColour: FormattingColours,
  bgColour: FormattingColours,
  bold = true,
  prefix = '',
  suffix = '',
) {
  const message = `${prefix}${chalk.bgHex(formattingColours[bgColour]).hex(formattingColours[fgColour])(`${string}${suffix}`)}`
  return bold ? chalk.bold(message) : message
}

export function formatRating(rating: Rating) {
  return `${formatString(rating.label, rating.fg, rating.bg)} ${rating.emoji}`
}

export function formatPercentageString(percentage: number) {
  for (const rating of ratings) {
    if (percentage >= rating.threshold) {
      return formatString(
        percentage.toFixed(2),
        rating.fg,
        rating.bg,
        true,
        '',
        '%',
      )
    }
  }

  // If the percentage is less than the lowest threshold, return it in white
  return formatString(percentage.toFixed(2), 'white', 'bgBlack', true, '', '%')
}
