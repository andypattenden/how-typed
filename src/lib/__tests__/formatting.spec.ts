import * as formatting from '../formatting.js'
import type { FormattingColours } from '../../@types/index.js'
import type { Rating } from '../ratings.js'

describe('formatting', () => {
  describe('formatString()', () => {
    it.each([
      [
        'moo',
        'green' as FormattingColours,
        'bgGreen' as FormattingColours,
        true,
        '',
        '',
      ],
      [
        'baa',
        'green' as FormattingColours,
        'bgGreen' as FormattingColours,
        true,
        'prefix',
        '',
      ],
      [
        'woof',
        'yellowGreen' as FormattingColours,
        'bgYellowGreen' as FormattingColours,
        true,
        '',
        'suffix',
      ],
      [
        'meow',
        'yellowGreen' as FormattingColours,
        'bgYellowGreen' as FormattingColours,
        true,
        'prefix',
        'suffix',
      ],
      [
        'roar',
        'yellowGreen' as FormattingColours,
        'bgYellowGreen' as FormattingColours,
        true,
        undefined,
        undefined,
      ],
    ])(
      "should return '%s' formatted",
      (
        string,
        fgColour: FormattingColours,
        bgColour: FormattingColours,
        isBold,
        prefix,
        suffix,
      ) => {
        const result = formatting.formatString(
          string,
          fgColour,
          bgColour,
          isBold,
          prefix,
          suffix,
        )
        expect(result).toMatchSnapshot()
      },
    )
  })

  describe('formatPercentageString()', () => {
    it.each([
      [95.5, '95.50%'],
      [85.25, '85.25%'],
      [65.75, '65.75%'],
      [55.5, '55.50%'],
      [35.1, '35.10%'],
      [25.99, '25.99%'],
    ])('for %s, it should return %s formatted', (percentage, expected) => {
      const result = formatting.formatPercentageString(percentage)
      expect(result).toContain(expected)
      expect(result).toMatchSnapshot()
    })
  })

  describe('formatRating()', () => {
    it.each([
      {
        threshold: 90,
        label: 'Excellent',
        fg: 'green',
        bg: 'bgGreen',
        emoji: '🚀',
      } as Rating,
      {
        threshold: 75,
        label: 'Great',
        fg: 'yellowGreen',
        bg: 'bgYellowGreen',
        emoji: '🎉',
      } as Rating,
      {
        threshold: 60,
        label: 'Good',
        fg: 'yellow',
        bg: 'bgYellow',
        emoji: '👍',
      } as Rating,
      {
        threshold: 50,
        label: 'Fair',
        fg: 'orange',
        bg: 'bgOrange',
        emoji: '🤏',
      } as Rating,
      {
        threshold: 30,
        label: 'Needs Improvement',
        fg: 'orangeRed',
        bg: 'bgOrangeRed',
        emoji: '😕',
      } as Rating,
      {
        threshold: 0,
        label: 'Poor',
        fg: 'red',
        bg: 'bgRed',
        emoji: '😞',
      } as Rating,
    ])(
      'for $label, it should return $label $emoji formatted',
      (rating: Rating) => {
        const result = formatting.formatRating(rating)
        expect(result).toContain(rating.label)
        expect(result).toContain(rating.emoji)
        expect(result).toMatchSnapshot()
      },
    )
  })
})
