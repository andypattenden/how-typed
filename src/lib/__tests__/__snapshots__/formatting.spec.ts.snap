// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`formatting > formatPercentageString() > for 25.99, it should return 25.99% formatted 1`] = `"[1m[48;2;51;0;0m[38;2;255;0;0m25.99%[39m[49m[22m"`;

exports[`formatting > formatPercentageString() > for 35.1, it should return 35.10% formatted 1`] = `"[1m[48;2;51;13;0m[38;2;255;69;0m35.10%[39m[49m[22m"`;

exports[`formatting > formatPercentageString() > for 55.5, it should return 55.50% formatted 1`] = `"[1m[48;2;51;26;0m[38;2;255;165;0m55.50%[39m[49m[22m"`;

exports[`formatting > formatPercentageString() > for 65.75, it should return 65.75% formatted 1`] = `"[1m[48;2;51;51;0m[38;2;255;255;0m65.75%[39m[49m[22m"`;

exports[`formatting > formatPercentageString() > for 85.25, it should return 85.25% formatted 1`] = `"[1m[48;2;51;77;0m[38;2;173;255;47m85.25%[39m[49m[22m"`;

exports[`formatting > formatPercentageString() > for 95.5, it should return 95.50% formatted 1`] = `"[1m[48;2;0;51;0m[38;2;0;255;0m95.50%[39m[49m[22m"`;

exports[`formatting > formatRating() > for 'Excellent', it should return 'Excellent' '🚀' formatted 1`] = `"[1m[48;2;0;51;0m[38;2;0;255;0mExcellent[39m[49m[22m 🚀"`;

exports[`formatting > formatRating() > for 'Fair', it should return 'Fair' '🤏' formatted 1`] = `"[1m[48;2;51;26;0m[38;2;255;165;0mFair[39m[49m[22m 🤏"`;

exports[`formatting > formatRating() > for 'Good', it should return 'Good' '👍' formatted 1`] = `"[1m[48;2;51;51;0m[38;2;255;255;0mGood[39m[49m[22m 👍"`;

exports[`formatting > formatRating() > for 'Great', it should return 'Great' '🎉' formatted 1`] = `"[1m[48;2;51;77;0m[38;2;173;255;47mGreat[39m[49m[22m 🎉"`;

exports[`formatting > formatRating() > for 'Needs Improvement', it should return 'Needs Improvement' '😕' formatted 1`] = `"[1m[48;2;51;13;0m[38;2;255;69;0mNeeds Improvement[39m[49m[22m 😕"`;

exports[`formatting > formatRating() > for 'Poor', it should return 'Poor' '😞' formatted 1`] = `"[1m[48;2;51;0;0m[38;2;255;0;0mPoor[39m[49m[22m 😞"`;

exports[`formatting > formatString() > should return 'baa' formatted 1`] = `"[1mprefix[48;2;0;51;0m[38;2;0;255;0mbaa[39m[49m[22m"`;

exports[`formatting > formatString() > should return 'meow' formatted 1`] = `"[1mprefix[48;2;51;77;0m[38;2;173;255;47mmeowsuffix[39m[49m[22m"`;

exports[`formatting > formatString() > should return 'moo' formatted 1`] = `"[1m[48;2;0;51;0m[38;2;0;255;0mmoo[39m[49m[22m"`;

exports[`formatting > formatString() > should return 'roar' formatted 1`] = `"[1m[48;2;51;77;0m[38;2;173;255;47mroar[39m[49m[22m"`;

exports[`formatting > formatString() > should return 'woof' formatted 1`] = `"[1m[48;2;51;77;0m[38;2;173;255;47mwoofsuffix[39m[49m[22m"`;
